### PR TITLE
core: resolve animation name in TraceElements gatherer

### PR DESCRIPTION
**Pull request by adamraine**
This change pertains to the non-composited animations audit: https://github.com/GoogleChrome/lighthouse/pull/11105

In addition to reporting nodes which cannot be composited we should also be reporting the display name of the animation if one exists.  For example, an animation with the keyframe effect below would have a display name of `example`.
```
@keyframes example {
    from {height: 100px};
    to {height: 200px};
}
```

This PR uses additional DevTools protocol functions `Animation.resolveAnimation` and `Runtime.getProperties` to resolve an animation ID in an animation trace event to the display name.

### CODE REVIEW

File: lighthouse-core/gather/gatherers/trace-elements.js [hunk 1]
```diff
@@ -136,22 +145,59 @@ class TraceElements extends Gatherer {
...
+   * @return {Promise<Array<TraceElementData>>}
    */
-  static getAnimatedElements(mainThreadEvents) {
-    const animatedElementIds = new Set(mainThreadEvents
-      .filter(e => e.name === 'Animation' && e.ph === 'b')
-      .map(e => this.getNodeIDFromTraceEvent(e)));
+  static async getAnimatedElements(passContext, mainThreadEvents) {
+    const driver = passContext.driver;
+
+    /** @type Map<number, string[]> */
```
Comment (patrickhulce) [inline lighthouse-core/gather/gatherers/trace-elements.js:155]: should this be...
```suggestion
    /** @type Map<number, Set<string>> */
```

File: lighthouse-core/gather/gatherers/trace-elements.js [hunk 2]
```diff
@@ -136,22 +145,59 @@ class TraceElements extends Gatherer {
...
+    const elementAnimations = new Map();
+    mainThreadEvents.filter(e => e.name === 'Animation' && e.ph === 'b')
+      .map(e => {
+        return {
+          nodeId: this.getNodeIDFromTraceEvent(e),
+          animationId: this.getAnimationIDFromTraceEvent(e),
+        };
+      })
+      .forEach(({nodeId, animationId}) => {
+        if (!nodeId || !animationId) return;
```
Comment (patrickhulce) [inline lighthouse-core/gather/gatherers/trace-elements.js:165]: can they ever be 0/empty?

File: lighthouse-core/gather/gatherers/trace-elements.js [hunk 3]
```diff
@@ -136,22 +145,59 @@ class TraceElements extends Gatherer {
...
+        const animationIds = elementAnimations.get(nodeId);
```
Comment (patrickhulce) [inline lighthouse-core/gather/gatherers/trace-elements.js:166]: just another tool for your toolbox, a pattern you'll frequently find throughout LH is something like

```js
const animationIds = elementAnimations.get(nodeId) || [];
animationIds.push(id);
elementAnimations.set(nodeId, animationIds);
```

but this works fine too :)

File: lighthouse-core/gather/gatherers/trace-elements.js [hunk 4]
```diff
@@ -203,12 +250,15 @@ class TraceElements extends Gatherer {
...
             ...response.result.value,
             score: backendNodeData[i].score,
+            animations: backendNodeData[i].animations,
             nodeId: backendNodeId,
           });
         }
       }
     }
 
+    driver.sendCommand('Animation.disable');
```
Comment (patrickhulce) [inline lighthouse-core/gather/gatherers/trace-elements.js:260]:
```suggestion
    await driver.sendCommand('Animation.disable');
```

File: lighthouse-core/gather/gatherers/trace-elements.js [hunk 5]
```diff
@@ -136,22 +145,59 @@ class TraceElements extends Gatherer {
...
     }
     return animatedElementData;
   }
 
+  /**
+   * @param {LH.Gatherer.PassContext} passContext
+   */
+  beforePass(passContext) {
+    passContext.driver.sendCommand('Animation.enable');
+  }
```
Comment (patrickhulce) [inline lighthouse-core/gather/gatherers/trace-elements.js:199]:
```suggestion
  async beforePass(passContext) {
    await passContext.driver.sendCommand('Animation.enable');
  }
```

File: lighthouse-core/gather/gatherers/trace-elements.js [hunk 6]
```diff
@@ -136,22 +145,59 @@ class TraceElements extends Gatherer {
...
+      });
 
     /** @type Array<TraceElementData> */
     const animatedElementData = [];
-    for (const nodeId of animatedElementIds) {
-      nodeId && animatedElementData.push({nodeId});
+    for (const [nodeId, animationIds] of elementAnimations) {
+      const animations = [];
+      for (const animationId of animationIds) {
+        const result = await driver.sendCommand('Animation.resolveAnimation', {animationId});
```
Comment (patrickhulce) [inline lighthouse-core/gather/gatherers/trace-elements.js:179]: does this ever fail if there was memory pressure and the old animation info got evicted?

if it works like the rest of the protocol I'm guessing it does and we should probably guard, maybe easier to create a `resolveAnimationId` function that returns `undefined` in a `catch` block for these sendCommand failure situations?

File: lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js [hunk 7]
```diff
@@ -215,6 +215,9 @@ module.exports = [
             width: 100,
             height: 100,
           },
+          animations: [
+            {name: 'anim'},
```
Comment (patrickhulce) [inline lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js:219]: cool so it is the name of the keyframe :)

File: lighthouse-core/test/gather/gatherers/trace-elements-test.js [hunk 8]
```diff
@@ -319,17 +319,40 @@ describe('Trace Elements gatherer - GetTopLayoutShiftElements', () => {
...
       makeAnimationTraceEvent('2', 5),
       makeAnimationTraceEvent('3', 6),
     ];
 
-    const result = TraceElementsGatherer.getAnimatedElements(traceEvents);
+    const result = await TraceElementsGatherer.getAnimatedElements({driver}, traceEvents);
     expect(result).toEqual([
-      {nodeId: 5},
-      {nodeId: 6},
+      {nodeId: 5, animations: [
```
Comment (patrickhulce) [inline lighthouse-core/test/gather/gatherers/trace-elements-test.js:349]: note to self: we should probably de-dupe the animation names on the audit side :)

File: lighthouse-core/gather/gatherers/trace-elements.js (hunk 1 above)
Comment (adamraine) [reply lighthouse-core/gather/gatherers/trace-elements.js:155]: Hmm I don't think we will ever encounter the same animation ID twice on one node. This is probably better to be safe.

File: lighthouse-core/gather/gatherers/trace-elements.js (hunk 2 above)
Comment (adamraine) [reply lighthouse-core/gather/gatherers/trace-elements.js:165]: No.  Node Id can't be 0 and animationId will be a string containing a number when defined.

File: lighthouse-core/gather/gatherers/trace-elements.js (hunk 6 above)
Comment (adamraine) [reply lighthouse-core/gather/gatherers/trace-elements.js:179]: This makes sense but should it also apply to resolving the node ID below?

File: lighthouse-core/gather/gatherers/trace-elements.js (hunk 1 above)
Comment (patrickhulce) [reply lighthouse-core/gather/gatherers/trace-elements.js]: will we encounter them across different nodes/should it be a single large set for resolution?

just feels odd to have a structure about IDs if everything is a tree that will never be repeated, but I guess if that's the way the trace events work we can stick to it.

File: lighthouse-core/gather/gatherers/trace-elements.js (hunk 2 above)
Comment (patrickhulce) [reply lighthouse-core/gather/gatherers/trace-elements.js:185]: cool 👍

File: lighthouse-core/gather/gatherers/trace-elements.js (hunk 6 above)
Comment (patrickhulce) [reply lighthouse-core/gather/gatherers/trace-elements.js]: yeah this refers to this entire inner loop section effectively

Comment (adamraine) [reply lighthouse-core/gather/gatherers/trace-elements.js]: The inner loop here is moved to the new function.  I was referring to the user of `driver.resolveNodeIdToObjectId` around line 184.

Comment (patrickhulce) [reply lighthouse-core/gather/gatherers/trace-elements.js]: Oh gotcha, `resolveNodeIdToObjectId` already does this internally
https://github.com/GoogleChrome/lighthouse/blob/cc5c79547388282f51d1f517890e17ca71c5e0de/lighthouse-core/gather/driver.js#L1227-L1235

File: lighthouse-core/gather/gatherers/trace-elements.js [hunk 9]
```diff
@@ -66,6 +75,28 @@ class TraceElements extends Gatherer {
...
+      const objectId = result.remoteObject.objectId;
+      if (!objectId) return undefined;
+      const response = await driver.sendCommand('Runtime.getProperties', {
+        objectId,
+      });
+      const nameProperty = response.result.find((property) => property.name === 'animationName');
+      const animationName = nameProperty && nameProperty.value && nameProperty.value.value;
+      return animationName;
+    } catch (err) {
+      return undefined;
```
Comment (patrickhulce) [inline lighthouse-core/gather/gatherers/trace-elements.js:96]:
```suggestion
      // Animation name is not mission critical information and can be evicted, so don't throw fatally if we can't find it.
      return undefined;
```

File: lighthouse-core/gather/gatherers/trace-elements.js (hunk 1 above)
Comment (adamraine) [reply lighthouse-core/gather/gatherers/trace-elements.js]: The animation IDs should be unique across all nodes.

**Pull request merged**